File: Gold이상문제_정리/문제집_1766.java
package Gold이상문제_정리;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.PriorityQueue;
import java.util.StringTokenizer;

public class 문제집_1766 {
	static int N, M;
	static int[] degree;
	static PriorityQueue<Integer> pq = new PriorityQueue<>();
	static ArrayList<Integer>[] adj;
	static StringBuilder str = new StringBuilder();
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());

		N = Integer.parseInt(st.nextToken());
		M = Integer.parseInt(st.nextToken());
		adj = new ArrayList[N+1];
		degree = new int[N+1];
		for (int i = 0; i < N+1; i++) {
			adj[i] = new ArrayList<>();
		}

		for (int i = 0; i < M; i++) {
			st = new StringTokenizer(br.readLine());
			int first = Integer.parseInt(st.nextToken());
			int next = Integer.parseInt(st.nextToken());
			adj[first].add(next);
			degree[next]++;
		}

		for (int i = 1; i < N+1; i++) {
			if (degree[i] == 0) {
				pq.add(i);
			}
		}
		while (!pq.isEmpty())
		{
			int now = pq.poll();
			str.append(now).append(" ");

			for(int next : adj[now]) {
				degree[next]--;
				if (degree[next] == 0) {
					pq.add(next);
				}
			}
		}
		System.out.println(str);
	}
}

File: 코틀린_코테스터디/week6/문제집_1766.kt
import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.*

private var N = 0
private var M:Int = 0
private var pq = PriorityQueue<Int>()
private lateinit var degree: IntArray
private lateinit var adj: Array<MutableList<Int>>
private var str = StringBuilder()

fun main() {
    val br = BufferedReader(InputStreamReader(System.`in`))
    var st = StringTokenizer(br.readLine())
    N = st.nextToken().toInt()
    M = st.nextToken().toInt()
    adj = Array(N + 1) { mutableListOf() }
    degree = IntArray(N + 1)
    for (i in 0 until N + 1) {
        adj[i] = ArrayList()
    }
    for (i in 0 until M) {
        st = StringTokenizer(br.readLine())
        val first = st.nextToken().toInt()
        val next = st.nextToken().toInt()
        adj[first].add(next)
        degree[next]++
    }
    for (i in 1 until N + 1) {
        if (degree[i] == 0) {
            pq.add(i)
        }
    }
    while (!pq.isEmpty()) {
        val now = pq.poll()
        str.append(now).append(" ")
        for (next in adj[now]) {
            degree[next]--
            if (degree[next] == 0) {
                pq.add(next)
            }
        }
    }
    println(str)
}